### DISCUSSION TITLE
route type?

### DISCUSSION THREAD

**Discussion by allotmentandy** [larastan/larastan#1429] [General]
Hi, loving phpstan / larastan today, got to level 8 already!! it is like a game now.

i have this code

```
14        $route = (\Route::current());
16        $data['brand'] = $route->wheres['brand'];
```

and this error

```
  16     Cannot access property $wheres on Illuminate\Routing\Route|null.
```

and ideas what the solution is?

i have written about it on my blog:

[https://allotmentandy.github.io/blog/2022-10-28-PHPStan-static-analysis/](https://allotmentandy.github.io/blog/2022-10-28-PHPStan-static-analysis/)

andy

**Comment by mad-briller** [comment]
there are situations in laravel where you will not have a route, such as in a custom `php artisan` command.
in such a case `Route::current()` will return null.

you can do a simple null check and throw if this is the case:

```php
$route = Route::current();

if ($route === null) {
    throw new UnexpectedValueException();
}

$data['brand'] = $route->wheres['brand'];
```

**Reply by allotmentandy** [reply]
thank you, that worked :)

**Comment by szepeviktor** [comment]
How about removing null?
```php
/** @var \Illuminate\Routing\Route  $route */
$route = (\Route::current());
```

**Reply by allotmentandy** [reply]
this works as well :)

**Reply by allotmentandy** [reply]
Viktor, how does that remove null? it is a comment

**Reply by mad-briller** [reply]
you are overriding the return type of the `current()` method with an explicit type, using the `@var` annotation
phpstan will trust what you tell it, over what it can work out for itself, if `treatPhpDocTypesAsCertain` is true ([which it is by default](https://phpstan.org/config-reference#treatphpdoctypesascertain))

this is fine if you are in a context where there will always be a route.. such as a route closure or controller perhaps, but if this is used in a service class that could be used in either location, `$route` could still actually be null and you've not protected yourself from it

**Reply by allotmentandy** [reply]
fantastic answer, thanks, it isnt code as such thou, it is just a note to say it is a route, in this example.

**Reply by mad-briller** [reply]
yeah you are telling phpstan "in this case $route can only be a Route, there is no situation where it can be null" which can range from fine - dangerous depending on the context

personally, i always steer towards handling the type rather than overriding phpstan's understanding of the type, as you never know what the future of the code holds;
for example after a refactor the `null` may become a problem but due to copy-paste the comment is still there, hiding that fact

**Reply by allotmentandy** [reply]
this makes sense

the comment from larastan was about line 16, is it possible to improve the comment

' 16     Cannot access property $wheres on Illuminate\Routing\Route|null.'

to say, check for null values

also the variable is the $route, not the $wheres part too.

**Reply by mad-briller** [reply]
that error is coming from phpstan rather than larastan, so changing the wording would have to be done there, and phpstan takes the backwards compatibility of it's error messages very seriously

i do agree it's a bit confusing at first, but you get used to its messaging because it is atleast consistent with how it describes things
